### ISSUE TITLE
Serializing `VecNormalize`

### ISSUE THREAD

**Issue by AdamGleave** [hill-a/stable-baselines#524]
**Describe the bug**
Using `VecNormalize` can speed up convergence of RL algorithms and reduce the need for hyperparameter tuning. The resulting policy will require normalized observations, making it essential to use `VecNormalize` as a wrapper when deploying the policy and to reload the statistics used for normalization.

Right now I have two pain points:
  1. Rather than just calling `model.save()`, I have to save the model and call `venv.save_running_average()`, and then when restoring create the environment and call `load_running_average()`. It'd be nice if there was a way to define model + transformations and save & restore them both in one line.
  2. `{save,load}_running_average` only save/restore the RMS statistics. In particular, they don't save `{norm,clip}_{obs,ret}`. Sometimes I don't want to normalize observations, just the returns. I have to save these parameters separately.

I don't have a good solution for (1). Plausibly `VecNormalize` should be a policy wrapper not an environment wrapper. This would also solve #200 I think, as the unnormalized observations would be placed in the replay buffer. I'm a little concerned this would make the transformations too opaque -- it's really useful to be able to visualize the observations that are actually going into your policy.

I think (2) can be addressed relatively simply by making `VecNormalize` picklable. We could then deprecate `{save,load}_running_average` which are just writing two pickle files anyway. The only thing that isn't serializable in `VecNormalize` is the underlying `venv`: we can simply omit that field, and add a `set_env` method.

I'd appreciate feedback on these ideas. If people are in favour I'm happy to make a PR to add `pickle` support to `VecNormalize` to fix (2).

**Comment by AdamGleave** [comment]
I mocked up an implementation of (2) in #525.

**Comment by araffin** [comment]
Hello,

For the point 1), I understand the motivation but this changes the current logic where the environment is decoupled from the agent. It gets even more complicated if you don't have only one wrapper.
We could do like DDPG does, and this would also solves #200 , so the normalization would be part of the tensorflow graph, but this complexify the code.

>it's really useful to be able to visualize the observations that are actually going into your policy.

I also agree on that point.

>I think (2) can be addressed relatively simply by making VecNormalize picklable

Good point for 2) ;)
